app/lib/request.rb

```diff
@@ -139,16 +139,23 @@ def body_with_limit(limit = 1.megabyte)
   class Socket < TCPSocket
     class << self
       def open(host, *args)
-        return super host, *args if thru_hidden_service? host
-        outer_e = nil
-        Addrinfo.foreach(host, nil, nil, :SOCK_STREAM) do |address|
-          begin
-            raise Mastodon::HostValidationError if PrivateAddressCheck.private_address? IPAddr.new(address.ip_address)
-            return super address.ip_address, *args
-          rescue => e
-            outer_e = e
+        return super(host, *args) if thru_hidden_service?(host)
+
+        outer_e   = nil
+        addresses = Addrinfo.getaddrinfo(host, nil, nil, :SOCK_STREAM).take(2)
```

Comment:
Why 2: AAAA, A is the most common use case. Spawning threads is not cheap, `Timeout.timeout` spawns threads, so we don't want to spawn more than 2 (one more than before)